src/go/pgo/distsys/state_server.go

```diff
@@ -0,0 +1,229 @@
+package distsys
+
+import (
+	"log"
+	"reflect"
+)
+
+func (t *NetworkRPC) GetRemote(args *GetRemoteArgs, reply *GetRemoteReply) error {
+	log.Printf("GetRemote(%v)\n", args)
+
+	errors := make(map[string]error)
+	redirects := make(map[string]string)
+	objects := make(map[string]RemoteObject)
+
+	for _, key := range args.Transaction.sorted() {
+		owner, ok := t.network.owners[key]
+		if !ok {
+			errors[key] = KeyNotFoundError(key)
+			continue
+		}
+
+		owner.RLock()
+		// log.Printf("locked key owner %v\n", key)
+
+		// disallow object access after redirecting, ensures locking order
+		// local requests will redirect to itself
+		if owner.isLocalhost(t.network) && len(redirects) == 0 {
+			var err error
+			var val interface{}
+
+			if args.Transaction.isExclusive(key) {
+				val, err = t.network.store.GetExclusive(key)
+			} else {
+				val, err = t.network.store.GetExclusive(key)
+				// TODO: for non-exclusive reads
+				// val, err = t.network.store.Get(key)
+			}
+
+			if err != nil {
+				errors[key] = err
+				continue
+			}
+
+			t.network.migration.OnGet(args.Host, key)
+
+			acquired := args.Transaction.isExclusive(key)
+			migrated := t.network.migration.MigrateTo(args.Host, key)
+			objects[key] = RemoteObject{
+				Value:    val,
+				Acquired: acquired,
+				Migrated: migrated,
+			}
+
+			// migration case will skip the remainder of the loop, not unlocking the owner
+			// until an ACK is recieved
+			if migrated {
+				/*
+					https://golang.org/src/sync/rwmutex.go
+
+					* If a goroutine holds a RWMutex for reading and another goroutine might call Lock,
+					* no goroutine should expect to be able to acquire a read lock until the initial
+					* read lock is released ...
+					* ... a blocked Lock call excludes new readers from acquiring the lock.
+
+					// A writer is pending.
+					74  		if atomic.AddInt32(&rw.readerWait, -1) == 0 {
+					75  			// The last reader unblocks the writer.
+					76  			runtime_Semrelease(&rw.writerSem, false)
+					77  		}
+
+					// Wait for active readers.
+					97  	if r != 0 && atomic.AddInt32(&rw.readerWait, r) != 0 {
+					98  		runtime_Semacquire(&rw.writerSem)
+					99  	}
+
+					thus if the current state is RLock, WLock will block until readerWait == 0;
+					however, this cannot happen until the current RLock is released (readerWait)
+
+					eg.
+					RLock() 	=> readerWait = 0
+					go WLock() 	=> readerWait = readerCount = 1 or more
+
+					while readerWait != 1
+					{ }			=> current state is not pending WLock
+					RUnlock() 	=> release the last reader, releasing the writerSem
+				*/
+				go owner.Lock()
+
+				for {
+					readers := reflect.ValueOf(owner.RWMutex).FieldByName("readerWait").Int()
```

Comment:
@fhackett Where was the comment above taken from? It would be great to chat about this in more details to see if we can avoid having to resort to reading private fields.

Reply:
wow - I heard about this change over IM but never got around to reading it properly.

To summarise: yes, we should fix that. I have a half-finished refactor lying around that was trying to address the issue, but it was a big change so Jnani "hacked" it to work with the demo.

The problem at hand is that it was a bad decision to make owner lockable in the first place - it initially caused a deadlock when implemented naively and this is a work-around for that. The correct way to handle this is to remove the owner.Lock critical section entirely and simply make reads/writes to owner data atomic. Then, we alter the local store so that when a thread reads that the current node is the owner but the variable migrates before we can lock the variable we just retry the read.